segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,413 @@
+ * current owner. Once the earlier owner received this notification, it stops writing to the segment.
+ *
+ */
+@Slf4j
+public class FileSystemStorage implements Storage {
+
+    //region members
+
+    private final FileSystemStorageConfig config;
+    private final ExecutorService executor;
+    private final AtomicBoolean closed;
+
+    //endregion
+
+    //region constructor
+
+    /**
+     * Creates a new instance of the FileSystemStorage class.
+     *
+     * @param config   The configuration to use.
+     * @param executor The executor to use for running async operations.
+     */
+    public FileSystemStorage(FileSystemStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.closed = new AtomicBoolean(false);
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+
+    /**
+     * Initialize is a no op here as we do not need a locking mechanism in case of file system write.
+     * @param containerEpoch The Container Epoch to initialize with (ignored here).
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+        this.ensureInitializedAndNotClosed();
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        return supplyAsync( streamSegmentName, () -> syncOpenRead(streamSegmentName));
+    }
+
+
+    @Override
+    public CompletableFuture<Integer> read( SegmentHandle handle,
+                                            long offset,
+                                            byte[] buffer,
+                                            int bufferOffset,
+                                            int length,
+                                            Duration timeout) {
+        return supplyAsync( handle.getSegmentName(), () -> syncRead(handle, offset, buffer, bufferOffset, length));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncGetStreamSegmentInfo(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncExists(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync( streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle,
```

Comment:
This was based on an earlier comment by @fpj. I will wait for you two to converge.